### ISSUE TITLE
Trigger-driven server-side PaperTrail

### ISSUE THREAD

**Issue by dark-panda** [paper-trail-gem/paper_trail#1170]
I have a Rails app that has a database that uses referential integrity and database triggers to handle some of its logic outside of the Rails application. When writes are performed outside of the Rails app, obviously PaperTrail won't track the changes, so I thought about how I would go about setting up a database-driven PaperTrail equivalent. Here's what I came up with:

https://gist.github.com/dark-panda/21483e488c9af0878a4b38ead01b4889

Basically, we have PaperTrail set up so that it disables all of the write callbacks and all of the writing is done via triggers. We still use PaperTrail for all read operations, for reifying old versions, etc., but the actual recording is done completely in the database. This works with triggers, referential integrity writes due to `on delete` and `on update` triggers, and works outside of the Rails app for any non-Rails applications we may have.

We can replicate the ability to selectively the PaperTrail writing on a per model and per request basis with some session-based configuration options that we can set and unset either through SQL or the Rails app. I don't think we follow the exact logic of PaperTrail 100% faithfully but this is more of a proof of concept -- we're considering rolling this out and not using it in production yet, but it has been working under development conditions pretty well. I haven't tested it yet at scale, because this would obviously change the overall structure where processing is done. It does take the load off of the application servers, but it centralizes all of the load in the database. This does result in some gains (less network traffic to the database, less processing Ruby-side, etc.) but may tie up the database more than desired. This would be something we'd have to instrument and is the sort of thing that is going to be very much use-case-specific to individual apps.

Is there any interest in furthering this technique, or is it useful to anyone? Perhaps this would be useful in a PaperTrail extension? I just figured I'd throw this up here to see if it's worth pursuing.

**Comment by dark-panda** [comment]
Yeah, the basic gist of things is that we used PaperTrail and built up some functionality around that, but found that our history had gaps in it from deletes and updates outside of our Rails app, which perhaps isn't super-common but we're probably not the only one to experience it. We didn't want to modify our app too much, so we wanted to keep on using PaperTrail, but needed to have the changes be captured regardless of their origin, so here we are.

I'll write up something more substantial and put it all into a gem when I get a chance. From all indications it looks like it will work out fine at our scale, so the idea probably valid overall.

**Comment by dark-panda** [comment]
Sorry to resurrect this issue from the grave, but we did eventually roll out our implementation of server-side paper trail and it's been up and running for a few months and it seems to be going well. I've extracted some of the relevant bits into a gist which you can find at the following hyperlink on the information super cyber highway pipes:

https://gist.github.com/dark-panda/21483e488c9af0878a4b38ead01b4889

**Comment by jaredbeck** [comment]
Neat! Thanks for sharing.
